2020/6/questionnaire/src/main.rs

```diff
@@ -0,0 +1,142 @@
+use std::collections::HashSet;
+use std::env;
+use std::fs::File;
+use std::io::{self, BufRead};
+
+
+fn main() -> Result<(), io::Error> {
+    let args: Vec<String> = env::args().collect();
+    let file = File::open(&args[1])?;
+    let reader = io::BufReader::new(file);
+
+    let mut vals = Vec::new();
+    let mut sum = 0;
+
+    for l in reader.lines() {
+        let line = l?;
+        if line.is_empty() {
+            sum += Group::new(&vals).num_all_yes();
+            vals.clear();
+        } else {
+            vals.push(line);
+        }
+    }
+
+    // Collect the final group
+    sum += Group::new(&vals).num_all_yes();
+
+    println!("The sum is {}", sum);
+
+    Ok(())
+}
+
+pub struct Group {
+    vals: Vec<String>,
+}
+
+impl Group {
+    pub fn new(vals: &[String]) -> Group {
```

Comment:
Unfortunately it's much trickier to make it so you can accept a `[String]` **or** `[str]` (to make those tests simpler).

It's doable, maybe with two generics: `I: Iterator<Target=T>, T: Into<String>` and then calling `vals.map(|v| v.into()).collect()`

That would mean an unnecessary copy of a `Vec<String>` argument though.

Ah the fun of Rust's polymorphism. I kinda like the lack of overloads to be honest but it's hard to get into the mindset. I think it's much more common to just expect your caller to get the arguments into the right type (let them build helpers if they need them) rather than building functions that accept anything.

Reply:
Yeah, did it for the first